File: Algorithms/SortingSeaching/RadixSort.py
def _sort_count(nums, mask):

    counts = [ 0 for _ in range(10) ]
    out = [ 0 for _ in nums ]

    for n in nums:
        counts[(n // mask) % 10] += 1

    for i in range(1, 10):
        counts[i] += counts[i - 1]

    for n in reversed(nums):
        out[counts[(n // mask) % 10] - 1] = n
        counts[(n // mask) % 10] -= 1

    nums[:] = out

def radix_sort(nums):

    m = max(nums)

    mask = 1

    while m / mask > 0:

        _sort_count(nums, mask)

        mask *= 10

def main():

    nums = [1, 52, 2, 25, 6, 10, 25, 200, 15, 12, 8, 2]

    print(nums)
    radix_sort(nums)
    print(nums)


if __name__ == '__main__':
    main()

File: Algorithms/SortingSeaching/RadixSort.js

let sortCount = (nums, mask) => {

    let counts = new Array(10);
    counts.fill(0);
    
    let out = new Array(nums.length);
    out.fill(0);

    for (let n of nums) {
        counts[Math.floor(n / mask) % 10]++;
    }

    for (let i = 1; i < 10; i++) {
        counts[i] += counts[i - 1];
    }

    for (let i = nums.length; i >= 0; i--) {
        out[counts[Math.floor(nums[i] / mask) % 10] - 1] = nums[i];
        counts[Math.floor(nums[i] / mask) % 10]--;
    }

    nums.splice(0, nums.length, ...out);

}

let radixSort = (nums) => {

    let m = Math.max(...nums)

    for (let mask = 1; Math.floor(m / mask) > 0; mask *= 10) {

        sortCount(nums, mask);

    }

}

let nums = [1, 52, 2, 25, 6, 10, 25, 200, 15, 12, 8, 2];

console.log(nums);

radixSort(nums);

console.log(nums);
